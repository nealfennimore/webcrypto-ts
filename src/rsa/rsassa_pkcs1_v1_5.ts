/**
 * Code related to RSASSA_PKCS1_v1_5
 * @module
 */

import * as params from "../params.js";
import * as proxy from "../proxy.js";
import { Alg as SHA } from "../sha/shared.js";
import {
    Alg,
    RsaShared,
    RsassaPkcs1V15CryptoKeyPair,
    RsassaPkcs1V15PrivCryptoKey,
    RsassaPkcs1V15ProxiedCryptoKeyPair,
    RsassaPkcs1V15ProxiedPrivCryptoKey,
    RsassaPkcs1V15ProxiedPubCryptoKey,
    RsassaPkcs1V15PubCryptoKey,
} from "./shared.js";

const handlers: proxy.ProxyKeyPairHandlers<
    RsassaPkcs1V15PrivCryptoKey,
    RsassaPkcs1V15PubCryptoKey
> = {
    privHandler: {
        get(target: RsassaPkcs1V15PrivCryptoKey, prop: string) {
            switch (prop) {
                case "self":
                    return target;
                case "sign":
                    return (data: BufferSource) => sign(target, data);
                case "exportKey":
                    return (format: KeyFormat) => exportKey(format, target);
            }

            return Reflect.get(target, prop);
        },
    },
    pubHandler: {
        get(target: RsassaPkcs1V15PubCryptoKey, prop: string) {
            switch (prop) {
                case "self":
                    return target;
                case "verify":
                    return (signature: BufferSource, data: BufferSource) =>
                        verify(target, signature, data);
                case "exportKey":
                    return (format: KeyFormat) => exportKey(format, target);
            }

            return Reflect.get(target, prop);
        },
    },
};

/**
 * Generate a new RSASSA_PKCS1_v1_5 keypair
 * @example
 * ```ts
 * const keyPair = await RSASSA_PKCS1_v1_5.generateKey();
 * ```
 */
export const generateKey = async (
    algorithm: Omit<params.EnforcedRsaHashedKeyGenParams, "name"> = {
        hash: SHA.Variant.SHA_512,
        modulusLength: 4096,
        publicExponent: new Uint8Array([0x01, 0x00, 0x01]),
    },
    extractable?: boolean,
    keyUsages?: KeyUsage[]
): Promise<RsassaPkcs1V15ProxiedCryptoKeyPair> => {
    const keyPair = (await RsaShared.generateKey(
        {
            ...algorithm,
            name: Alg.Variant.RSASSA_PKCS1_v1_5,
        },
        extractable,
        keyUsages
    )) as RsassaPkcs1V15CryptoKeyPair;
    return proxy.proxifyKeyPair<
        RsassaPkcs1V15CryptoKeyPair,
        RsassaPkcs1V15PrivCryptoKey,
        RsassaPkcs1V15ProxiedPrivCryptoKey,
        RsassaPkcs1V15PubCryptoKey,
        RsassaPkcs1V15ProxiedPubCryptoKey
    >(handlers)(keyPair);
};

/**
 * Generate a new RSASSA_PKCS1_v1_5 keypair
 * @alias generateKey
 * @example
 * ```ts
 * const keyPair = await RSASSA_PKCS1_v1_5.generateKeyPair();
 * ```
 */
export const generateKeyPair = generateKey;

/**
 * Import an RSASSA_PKCS1_v1_5 public or private key
 * @example
 * ```ts
 * const key = await RSASSA_PKCS1_v1_5.importKey("jwk", pubKey, { hash: "SHA-512" }, true, ['verify']);
 * ```
 */
export const importKey = async (
    format: KeyFormat,
    key: BufferSource | JsonWebKey,
    algorithm: Omit<params.EnforcedRsaHashedImportParams, "name">,
    extractable?: boolean,
    keyUsages?: KeyUsage[]
): Promise<
    RsassaPkcs1V15ProxiedPubCryptoKey | RsassaPkcs1V15ProxiedPrivCryptoKey
> => {
    const importedKey = await RsaShared.importKey(
        format,
        key,
        { ...algorithm, name: Alg.Variant.RSASSA_PKCS1_v1_5 },
        extractable,
        keyUsages
    );

    if (importedKey.type === "private") {
        return proxy.proxifyKey<
            RsassaPkcs1V15PrivCryptoKey,
            RsassaPkcs1V15ProxiedPrivCryptoKey
        >(handlers.privHandler)(importedKey as RsassaPkcs1V15PrivCryptoKey);
    } else {
        return proxy.proxifyKey<
            RsassaPkcs1V15PubCryptoKey,
            RsassaPkcs1V15ProxiedPubCryptoKey
        >(handlers.pubHandler)(importedKey as RsassaPkcs1V15PubCryptoKey);
    }
};

/**
 * Export an RSASSA_PKCS1_v1_5 public or private key
 * @example
 * ```ts
 * const pubKeyJwk = await RSASSA_PKCS1_v1_5.importKey("jwk", keyPair.publicKey.self);
 * ```
 * @example
 * ```ts
 * const pubKeyJwk = await keyPair.publicKey.importKey("jwk");
 * ```
 */
export const exportKey = async (
    format: KeyFormat,
    key: RsassaPkcs1V15PubCryptoKey | RsassaPkcs1V15PrivCryptoKey
) => RsaShared.exportKey(format, key);

/**
 * Sign a given payload
 * @example
 * ```ts
 * const message = new TextEncoder().encode("a message");
 * const signature = await RSASSA_PKCS1_v1_5.sign(keyPair.privateKey.self, message);
 * ```
 * @example
 * ```ts
 * const message = new TextEncoder().encode("a message");
 * const signature = await keyPair.privateKey.sign(message);
 * ```
 */
export const sign = async (
    key: RsassaPkcs1V15PrivCryptoKey,
    data: BufferSource
) =>
    await RsaShared.sign(
        {
            name: Alg.Variant.RSASSA_PKCS1_v1_5,
        },
        key,
        data
    );

/**
 * Verify a given signature
 * @example
 * ```ts
 * const message = new TextEncoder().encode("a message");
 * const isVerified = await RSASSA_PKCS1_v1_5.verify(keyPair.publicKey.self, signature, message);
 * ```
 * @example
 * ```ts
 * const message = new TextEncoder().encode("a message");
 * const isVerified = await keyPair.publicKey.verify( signature, message);
 * ```
 */
export const verify = async (
    key: RsassaPkcs1V15PubCryptoKey,
    signature: BufferSource,
    data: BufferSource
) =>
    await RsaShared.verify(
        {
            name: Alg.Variant.RSASSA_PKCS1_v1_5,
        },
        key,
        signature,
        data
    );
