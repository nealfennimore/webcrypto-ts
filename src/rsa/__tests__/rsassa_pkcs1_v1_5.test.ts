import { RSASSA_PKCS1_v1_5 } from "../rsassa_pkcs1_v1_5";

describe("RSASSA_PKCS1_v1_5", () => {
    it("should generate key", async () => {
        const key = await RSASSA_PKCS1_v1_5.generateKey();
        expect(key).toMatchSnapshot();
    });

    it("should import and export key", async () => {
        const keyPair = await RSASSA_PKCS1_v1_5.generateKey();

        let jwk = await RSASSA_PKCS1_v1_5.exportKey("jwk", keyPair.publicKey);
        const pubKey = await RSASSA_PKCS1_v1_5.importKey(
            "jwk",
            "SHA-512",
            jwk,
            true,
            ["verify"]
        );

        expect(jwk).toEqual(await RSASSA_PKCS1_v1_5.exportKey("jwk", pubKey));

        jwk = await RSASSA_PKCS1_v1_5.exportKey("jwk", keyPair.privateKey);
        const privKey = await RSASSA_PKCS1_v1_5.importKey(
            "jwk",
            "SHA-512",
            jwk,
            true,
            ["sign"]
        );

        expect(jwk).toEqual(await RSASSA_PKCS1_v1_5.exportKey("jwk", privKey));
    });
    it("should sign and verify", async () => {
        const keyPair = await RSASSA_PKCS1_v1_5.generateKey();
        const text = encode("a message");
        const signature = await RSASSA_PKCS1_v1_5.sign(
            keyPair.privateKey,
            text
        );
        expect(
            await RSASSA_PKCS1_v1_5.verify(keyPair.publicKey, signature, text)
        ).toEqual(true);
    });
});
