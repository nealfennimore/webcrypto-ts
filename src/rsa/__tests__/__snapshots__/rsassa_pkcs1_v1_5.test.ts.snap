// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RSASSA_PKCS1_v1_5 Original should generate key 1`] = `
Object {
  "privateKey": CryptoKey {
    Symbol(kKeyObject): PrivateKeyObject {
      Symbol(kKeyType): "private",
    },
    Symbol(kAlgorithm): Object {
      "hash": Object {
        "name": "SHA-512",
      },
      "modulusLength": 4096,
      "name": "RSASSA-PKCS1-v1_5",
      "publicExponent": Uint8Array [
        1,
        0,
        1,
      ],
    },
    Symbol(kExtractable): true,
    Symbol(kKeyUsages): Array [
      "sign",
    ],
  },
  "publicKey": CryptoKey {
    Symbol(kKeyObject): PublicKeyObject {
      Symbol(kKeyType): "public",
    },
    Symbol(kAlgorithm): Object {
      "hash": Object {
        "name": "SHA-512",
      },
      "modulusLength": 4096,
      "name": "RSASSA-PKCS1-v1_5",
      "publicExponent": Uint8Array [
        1,
        0,
        1,
      ],
    },
    Symbol(kExtractable): true,
    Symbol(kKeyUsages): Array [
      "verify",
    ],
  },
}
`;

exports[`RSASSA_PKCS1_v1_5 Proxied should generate key 1`] = `
Object {
  "privateKey": CryptoKey {
    Symbol(kKeyObject): PrivateKeyObject {
      Symbol(kKeyType): "private",
    },
    Symbol(kAlgorithm): Object {
      "hash": Object {
        "name": "SHA-512",
      },
      "modulusLength": 4096,
      "name": "RSASSA-PKCS1-v1_5",
      "publicExponent": Uint8Array [
        1,
        0,
        1,
      ],
    },
    Symbol(kExtractable): true,
    Symbol(kKeyUsages): Array [
      "sign",
    ],
  },
  "publicKey": CryptoKey {
    Symbol(kKeyObject): PublicKeyObject {
      Symbol(kKeyType): "public",
    },
    Symbol(kAlgorithm): Object {
      "hash": Object {
        "name": "SHA-512",
      },
      "modulusLength": 4096,
      "name": "RSASSA-PKCS1-v1_5",
      "publicExponent": Uint8Array [
        1,
        0,
        1,
      ],
    },
    Symbol(kExtractable): true,
    Symbol(kKeyUsages): Array [
      "verify",
    ],
  },
}
`;

exports[`RSASSA_PKCS1_v1_5 should generate key 1`] = `
Object {
  "privateKey": CryptoKey {
    Symbol(kKeyObject): PrivateKeyObject {
      Symbol(kKeyType): "private",
    },
    Symbol(kAlgorithm): Object {
      "hash": Object {
        "name": "SHA-512",
      },
      "modulusLength": 4096,
      "name": "RSASSA-PKCS1-v1_5",
      "publicExponent": Uint8Array [
        1,
        0,
        1,
      ],
    },
    Symbol(kExtractable): true,
    Symbol(kKeyUsages): Array [
      "sign",
    ],
  },
  "publicKey": CryptoKey {
    Symbol(kKeyObject): PublicKeyObject {
      Symbol(kKeyType): "public",
    },
    Symbol(kAlgorithm): Object {
      "hash": Object {
        "name": "SHA-512",
      },
      "modulusLength": 4096,
      "name": "RSASSA-PKCS1-v1_5",
      "publicExponent": Uint8Array [
        1,
        0,
        1,
      ],
    },
    Symbol(kExtractable): true,
    Symbol(kKeyUsages): Array [
      "verify",
    ],
  },
}
`;
